packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -763,7 +764,19 @@ function applyKey(
             }
         }
 
-        element.directives.push(ast.keyDirective(keyAttribute.value, keyAttribute.location));
+        const parentIsIteration = !!(forOfParent || forEachParent);
+
+        if (!parentIsIteration) {
+            ctx.warnAtLocation(
+                ParserDiagnostics.KEY_SHOULD_BE_CHILD_OF_ITERATION,
+                ast.sourceLocation(parse5ElmLocation),
+                [tag]
+            );
+        }
+
+        element.directives.push(
+            ast.keyDirective(keyAttribute.value, keyAttribute.location, parentIsIteration)
+        );
```

Comment:
I don't think it makes sense to keep the `key` directive as part of the AST if it doesn't carry any semantic.

Reply:
You're right. Fixed.